Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -0,0 +1,150 @@
+Here `p` is a prime number.
+-/
+
+namespace Nat
+
+/-- `primesBelow n` is the set of primes less than `n` as a finset. -/
+def primesBelow (n : ℕ) : Finset ℕ := (Finset.range n).filter (fun p ↦ p.Prime)
+
+@[simp]
+lemma primesBelow_zero : primesBelow 0 = ∅ := by
+  rw [primesBelow, Finset.range_zero, Finset.filter_empty]
+
+lemma prime_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p.Prime :=
+  (Finset.mem_filter.mp h).2
+
+lemma lt_of_mem_primesBelow {p n : ℕ} (h : p ∈ n.primesBelow) : p < n :=
+  Finset.mem_range.mp <| Finset.mem_of_mem_filter p h
+
+lemma primesBelow_succ (n : ℕ) :
+    primesBelow n.succ = if n.Prime then insert n (primesBelow n) else primesBelow n := by
+  rw [primesBelow, primesBelow, Finset.range_succ, Finset.filter_insert]
+
+lemma not_mem_primesBelow (n : ℕ) : n ∉ primesBelow n :=
+  fun hn ↦ (lt_of_mem_primesBelow hn).false
+
+/-- `smoothNumbers n` is the set of *`n`-smooth positive natural numbers*, i.e., the
+positive natural numbers all of whose prime factors are less than `n`. -/
+def smoothNumbers (n : ℕ) : Set ℕ := {m | m ≠ 0 ∧ ∀ p ∈ factors m, p < n}
+
+lemma mem_smoothNumbers {n m : ℕ} : m ∈ smoothNumbers n ↔ m ≠ 0 ∧ ∀ p ∈ factors m, p < n :=
+  Iff.rfl
+
+@[simp]
+lemma smoothNumbers_zero : smoothNumbers 0 = {1} := by
+  ext m
+  rw [Set.mem_singleton_iff, mem_smoothNumbers]
+  simp_rw [not_lt_zero]
+  rw [← List.eq_nil_iff_forall_not_mem, factors_eq_nil, and_or_left, not_and_self_iff, false_or,
+    ne_and_eq_iff_right zero_ne_one]
+
+/-- The product of the prime factors of `n` that are less than `N` is an `N`-smooth number. -/
+lemma prod_mem_smoothNumbers (n N : ℕ) : (n.factors.filter (· < N)).prod ∈ smoothNumbers N := by
+  have h₀ : (n.factors.filter (· < N)).prod ≠ 0 :=
+    List.prod_ne_zero fun h ↦ (pos_of_mem_factors (List.mem_of_mem_filter h)).false
+  refine ⟨h₀, fun p hp ↦ ?_⟩
+  obtain ⟨H₁, H₂⟩ := (mem_factors h₀).mp hp
+  simpa only [decide_eq_true_eq] using List.of_mem_filter <| mem_list_primes_of_dvd_prod H₁.prime
+    (fun _ hq ↦ (prime_of_mem_factors (List.mem_of_mem_filter hq)).prime) H₂
+
+/-- The sets of `N`-smooth and of `(N+1)`-smooth numbers are the same when `N` is not prime.
+See `Nat.equivProdNatSmoothNumbers` for when `N` is prime. -/
+lemma smoothNumbers_succ {N : ℕ} (hN : ¬ N.Prime) : N.succ.smoothNumbers = N.smoothNumbers := by
+  ext m
+  refine ⟨fun hm ↦ ⟨hm.1, fun p hp ↦ ?_⟩,
+         fun hm ↦ ⟨hm.1, fun p hp ↦ (hm.2 p hp).trans <| lt.base N⟩⟩
+  exact lt_of_le_of_ne (lt_succ.mp <| hm.2 p hp) fun h ↦ hN <| h ▸ prime_of_mem_factors hp
+
+@[simp] lemma smoothNumbers_one : smoothNumbers 1 = {1} := by simp [smoothNumbers_succ]
+
+@[gcongr] lemma smoothNumbers_mono {N M : ℕ} (hNM : N ≤ M) : N.smoothNumbers ⊆ M.smoothNumbers :=
+  fun _ hx ↦ ⟨hx.1, fun p hp => (hx.2 p hp).trans_le hNM⟩
+
+/-- The non-zero non-`N`-smooth numbers are `≥ N`. -/
+lemma smoothNumbers_compl (N : ℕ) : (N.smoothNumbers)ᶜ \ {0} ⊆ {n | N ≤ n} := by
```

Comment:
I thought 0 isn't smooth anyways?

Reply:
Exactly, so 0 is in the complement, so without removing it, the statement would be false.